org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java

```diff
@@ -0,0 +1,655 @@
+  }
+
+  private static JarEntryFile findJarFile(JarEntryDirectory directory, String path) {
+    for (IJarEntryResource children : directory.getChildren()) {
+      if (children.isFile() && children.getFullPath().toOSString().equals(path)) {
+        return (JarEntryFile) children;
+      }
+      if (!children.isFile()) {
+        JarEntryFile file = findJarFile((JarEntryDirectory) children, path);
+        if (file != null) {
+          return file;
+        }
+      }
+    }
+    return null;
+  }
+
+  private static IPackageFragmentRoot getPackageFragmentRoot(
+      IJavaProject project, String id, IProgressMonitor pm) throws JavaModelException {
+    IPackageFragmentRoot[] roots = project.getAllPackageFragmentRoots();
+    IPackageFragmentRoot packageFragmentRoot = null;
+    for (IPackageFragmentRoot root : roots) {
+      if (pm.isCanceled()) {
+        throw new OperationCanceledException();
+      }
+      if (root.getHandleIdentifier().equals(id)) {
+        packageFragmentRoot = root;
+        break;
+      }
+    }
+    return packageFragmentRoot;
+  }
+
+  private static String readFileContent(JarEntryFile file) {
+    try (InputStream stream = (file.getContents())) {
+      return IOUtils.toString(stream, "UTF-8");
+    } catch (IOException | CoreException e) {
+      JavaLanguageServerPlugin.logException("Can't read file content: " + file.getFullPath(), e);
+    }
+    return null;
+  }
+
+  private static Object[] findJarDirectoryChildren(JarEntryDirectory directory, String path) {
+    String directoryPath = directory.getFullPath().toOSString();
+    if (directoryPath.equals(path)) {
+      return directory.getChildren();
+    }
+    if (path.startsWith(directoryPath)) {
+      for (IJarEntryResource resource : directory.getChildren()) {
+        String childrenPath = resource.getFullPath().toOSString();
+        if (childrenPath.equals(path)) {
+          return resource.getChildren();
+        }
+        if (path.startsWith(childrenPath) && resource instanceof JarEntryDirectory) {
+          findJarDirectoryChildren((JarEntryDirectory) resource, path);
+        }
+      }
+    }
+    return null;
+  }
+
+  private static Object[] getPackageFragmentRootContent(
+      IPackageFragmentRoot root, IProgressMonitor pm) throws JavaModelException {
+    ArrayList<Object> result = new ArrayList<>();
+    getHierarchicalPackageChildren(root, null, result, pm);
+    Object[] nonJavaResources = root.getNonJavaResources();
+    Collections.addAll(result, nonJavaResources);
+    return result.toArray();
+  }
+
+  /**
+   * Returns the hierarchical packages inside a given fragment or root.
+   *
+   * @param parent the parent package fragment root
+   * @param fragment the package to get the children for or 'null' to get the children of the root
+   * @param result Collection where the resulting elements are added
+   * @throws JavaModelException if fetching the children fails
+   */
+  private static void getHierarchicalPackageChildren(
```

Comment:
If you pass in a collection to be filled, it would be nicer to call the method "collect...."

Reply:
fixed